### PR TITLE
Charly/leaderelection configmap

### PR DESCRIPTION
**Pull request by CharlyF**
### What does this PR do?

Moving the leader election to rely on configmaps a la Agent 5.

### Motivation

So that the helm chart's RBAC does not need to be modified.

### Testings

```
ubuntu@ci-charly:/go/src/github.com/DataDog/datadog-agent$ go test -tags "kubeapiserver docker" ./test/integration/util/leaderelection/ -v
=== RUN   TestSuiteAPIServer
[...]
=== RUN   TestSuiteAPIServer/TestLeaderElectionMulti
1519189667930732751 [Debug] HolderIdentity is "test-multi-0"
1519189667930786601 [Debug] LeaderLeaseDuration is 1m0s
1519189667938717163 [Debug] Current registered leader is ""
1519189667938793910 [Debug] Leader Engine successfully initialized
1519189667938820586 [Debug] HolderIdentity is "test-multi-1"
1519189667938922149 [Debug] LeaderLeaseDuration is 1m0s
=== RUN   TestSuiteAPIServer/TestLeaderElectionMulti/test-multi-0-0
1519189667951043647 [Debug] Current registered leader is ""
1519189667951077399 [Debug] Leader Engine successfully initialized
1519189667951740108 [Info] Starting Leader Election process for "test-multi-0" ...
1519189667957215943 [Info] Currently new leader "test-multi-0"
1519189667957261995 [Info] Leading as "test-multi-0" ...
=== RUN   TestSuiteAPIServer/TestLeaderElectionMulti/test-multi-1-1
1519189668461546629 [Info] Leader Election run, current leader is "test-multi-0"
1519189669464210462 [Info] Starting Leader Election process for "test-multi-1" ...
1519189669466600576 [Info] Currently new leader "test-multi-0"
1519189669973945940 [Info] Leader Election run, current leader is "test-multi-0"
1519189671982698726 [Info] leader is test-multi-0
--- PASS: TestSuiteAPIServer (20.23s)
    --- PASS: TestSuiteAPIServer/TestLeaderElectionMulti (4.57s)
        --- PASS: TestSuiteAPIServer/TestLeaderElectionMulti/test-multi-0-0 (0.51s)
        --- PASS: TestSuiteAPIServer/TestLeaderElectionMulti/test-multi-1-1 (1.51s)
PASS
ok  	github.com/DataDog/datadog-agent/test/integration/util/leaderelection	20.288s
```

Testing in real:
CM is created and has the correct metadata.
```
ubuntu@ci-charly:~$ kubectl describe cm datadog-leader-election
Name:         datadog-leader-election
Namespace:    default
Labels:       <none>
Annotations:  control-plane.alpha.kubernetes.io/leader={"holderIdentity":"datadog-agent-7bc6b4b775-82xhq","leaseDurationSeconds":60,"acquireTime":"2018-02-21T04:54:53Z","renewTime":"2018-02-21T04:56:53Z","leaderTra...

Data
====
Events:  <none>
```

Node1 (leader):
```
2018-02-21 04:54:53 UTC | DEBUG | (leaderelection_engine.go:68 in newElection) | Current registered leader is ""
2018-02-21 04:54:53 UTC | DEBUG | (leaderelection.go:132 in init) | Leader Engine successfully initialized
2018-02-21 04:54:53 UTC | INFO | (leaderelection.go:137 in startLeaderElection) | Starting Leader Election process for "datadog-agent-7bc6b4b775-82xhq" ...
2018-02-21 04:54:53 UTC | INFO | (leaderelection_engine.go:75 in func1) | Currently new leader "datadog-agent-7bc6b4b775-82xhq"
2018-02-21 04:54:53 UTC | INFO | (leaderelection_engine.go:78 in func2) | Leading as "datadog-agent-7bc6b4b775-82xhq" ...
2018-02-21 04:54:54 UTC | INFO | (leaderelection.go:155 in EnsureLeaderElectionRuns) | Leader Election run, current leader is "datadog-agent-7bc6b4b775-82xhq"
[...]
2018-02-21 04:59:23 UTC | INFO | (leaderelection.go:155 in EnsureLeaderElectionRuns) | Leader Election run, current leader is "datadog-agent-7bc6b4b775-82xhq"
```

Node2 (follower):
```
2018-02-21 04:59:14 UTC | INFO | (leaderelection.go:155 in EnsureLeaderElectionRuns) | Leader Election run, current leader is "datadog-agent-7bc6b4b775-82xhq"
2018-02-21 04:59:14 UTC | DEBUG | (kubernetes_apiserver.go:161 in runLeaderElection) | Leader is "datadog-agent-7bc6b4b775-82xhq". datadog-agent-1-7bc6b4b775-fcjcx will not run Kubernetes cluster related checks and collecting events
```

### CODE REVIEW

**Comment by CharlyF** [comment]
Still need to look at the `Current registered leader is ""` even though that does not seem to be a problem.

**Review by JulienBalestra — DISMISSED** [review]
LGTM

**Comment by CharlyF** [comment]
rebased on master to have the glog fix. For some reason GH did not mention that this branch was outdated.

File: Dockerfiles/manifests/rbac/clusterrole.yaml [hunk 1]
```diff
@@ -28,8 +28,9 @@ rules:
 - apiGroups:
   - ""
   resources:
-  - endpoints
+  - configmaps
   verbs:
   - get
   - update
   - create
+  - list
```
Comment (mfpierre) [inline Dockerfiles/manifests/rbac/clusterrole.yaml]: why do we need this additional `list` verb compared to the `endpoints`?

File: test/integration/util/leaderelection/leaderelection_test.go [hunk 2]
```diff
@@ -105,10 +105,10 @@ func (suite *apiserverSuite) waitForLeaderName(le *leaderelection.LeaderEngine)
 		case <-tick.C:
 			leaderName = le.CurrentLeaderName()
 			if leaderName == le.HolderIdentity {
-				log.Infof("leader is %s", leaderName)
+				log.Infof("waiting for leader: Leader is %q", leaderName)
```
Comment (mfpierre) [inline test/integration/util/leaderelection/leaderelection_test.go]: usually you don't capitalize after a colon

File: test/integration/util/leaderelection/leaderelection_test.go [hunk 3]
```diff
@@ -105,10 +105,10 @@ func (suite *apiserverSuite) waitForLeaderName(le *leaderelection.LeaderEngine)
...
 				return
 			}
-			log.Infof("leader is %q", leaderName)
+			log.Infof("waiting for leader: Leader is %q", leaderName)
```
Comment (mfpierre) [inline test/integration/util/leaderelection/leaderelection_test.go]: ditto

File: pkg/util/kubernetes/apiserver/leaderelection/leaderelection_engine.go [hunk 4]
```diff
@@ -19,52 +19,54 @@ import (
...
+	configMap, err := le.coreClient.ConfigMaps(namespace).Get(electionId, metav1.GetOptions{})
 	if err != nil {
 		return "", nil, err
 	}
 
-	val, found := endpoint.Annotations[rl.LeaderElectionRecordAnnotationKey]
+	val, found := configMap.Annotations[rl.LeaderElectionRecordAnnotationKey]
 	if !found {
-		return "", endpoint, nil
+		log.Infof("The configmap/%s in the namespace %s doesn't have the annotation %q: no-one is leading yet", electionId, namespace, rl.LeaderElectionRecordAnnotationKey)
```
Comment (mfpierre) [inline pkg/util/kubernetes/apiserver/leaderelection/leaderelection_engine.go]: Should it be `info` or `debug`??

**Review by mfpierre — DISMISSED** [review]
Looking good 👍 

Left a couple of remarks but I'm good with the merge

File: Dockerfiles/manifests/rbac/clusterrole.yaml (hunk 1 above)
Comment (CharlyF) [reply Dockerfiles/manifests/rbac/clusterrole.yaml]: @mfpierre thank you for bringing this up.
For the Service Mapper (and other functions like the checkResourcesAuth) we need to list the Endpoints. This verb is above in the first section of the role. We never used to need to list the configmaps.
Now that the Leader Election relies on configmaps we need to List them.
We still need to List the endpoints though, which is why the verb was added.

**Comment by CharlyF** [comment]
Changes look good to me, thanks for the update!

**Review by mfpierre — APPROVED** [review]
:shipit:

**Pull request merged**